### PR TITLE
Simplify Baggage handling in the OpenTracing Shim layer.

### LINKED ISSUE #2137: OpenTracing layer support for Baggage
In OpenTracing, `SpanContext` and `Baggage` are part of the same object, whereas in OpenTelemetry they are different objects. This would be straightforward and trivial to support **if** the **OpenTracing Shim** couldn't be used along the OpenTelemetry API in the same code base -- which, however, has been deemed important to support gradual adoption in large codebases. 

This currently is supported in the **OpenTracing Shim** by associating any OpenTelemetry `Span` with a given `SpanContext` + `Baggage` combo at all times, in order to see changes reflected in different threads at all times. A typical use case is

```java
// Main thread, OTel API
io.opentelemetry.trace.Span span1 = ...; // Create OpenTelemetry Span alone
propagateAsCurrentSpanToThreads(span1);

// A thread consuming the the propagated span1 through the OT Shim
openTracingTracer.activeSpan().setBaggageItem("hello", "foo");

// Another thread consuming the propagated span1 through the OT Shim too
openTracingTracer.activeSpan().getBaggageItem("hello"); // returns foo
```

This is supported in [Python easily](https://github.com/open-telemetry/opentelemetry-python/compare/main...carlosalberto:ot_shim_baggage_multithread?expand=1) by decorating a pair of values in the OpenTelemetry `Span` itself, whereas in Java (and similar languages) it requires a global map with a global read-write lock, which _may_ still impact the performance greatly. 

### Alternative 1

Completely discard support for Baggage, providing the related functionality as no-op, given:

* Its complicated handling (for some languages) when the OT Shim is used along the OTel API.
* Its potential lack of common usage in OpenTracing
* Additionally: its heavy memory footprint (as `SpanContext` baggage values are **copied** from the the parent `Span`).

### Alternative 2

Provide a _best-hope_ design, in a relatively simplistic fashion, in order to handle common cases (such as an OpenTracing `Span` being created, have its baggage set, and have its baggage propagated).

* Have `SpanContext` + `Baggage` be part of the `Span` **Shim** object (hence, no need for global maps/locks), and propagate this object itself (along its current `Baggage`). This would allow:
  * OTel API usage: simply consume the propagated `Baggage`. 
  * OT Shim usage: have different threads **update** the `Baggage` value as needed.
* If the current `Context` has no `Span` **Shim** object, all the OT Shim Baggage operations are no-op.

This WOULD NOT satisfy the expected OT Handling, but it would offer limited support though.

cc @yurishkuro @tedsuo

### PR DESCRIPTION
**Pull request by carlosalberto**
Baggage is NOT directly associated with an OpenTelemetry `Span` directly anymore, but with the **OpenTracing Shim layer**, which greatly simplifies its handling (no more global dictionary or extra field in Spans).

This comes with the recommendation of NOT consuming the OpenTelemetry API and the OpenTracing Shim the codebase, as the `Span`'s associated `Baggage` may be lost during in-process or inter-process propagation.

I have a [prototype](https://github.com/open-telemetry/opentelemetry-java/compare/main...carlosalberto:otshim-spanshim-lock?expand=1) for this in Java.

I'd added thread-safety requirements for `Baggage` consumption, even if it doesn't exist in the original OpenTracing Specification, as those calls are thread-safe anyway in most of (all?) OpenTracing compliant tracers.

Fixes #2137

### CODE REVIEW

**Comment by carlosalberto** [comment]
Ping @yurishkuro

File: specification/compatibility/opentracing.md [hunk 1]
```diff
@@ -44,14 +44,14 @@ upstream OpenTelemetry API.
...
 
-The OpenTracing Shim and the OpenTelemetry API/SDK are expected to be consumed
-simultaneously in a running service, in order to ease migration from the former
-to the latter.
-
 Semantic convention mapping SHOULD NOT be performed, with the
 exception of error mapping, as described in the [Set Tag](#set-tag) and
 [Log](#log) sections.
 
+It is not recommended the OpenTracing Shim and the OpenTelemetry API are
```
Comment (yurishkuro) [inline specification/compatibility/opentracing.md:51]: Isn't this too strong? I think this recommendation only stands when conditioned on "if you want to pass baggage through the application, then don't mix shim and core API".

**Review by yurishkuro — APPROVED** [review]

File: specification/compatibility/opentracing.md (hunk 1 above)
Comment (carlosalberto) [reply specification/compatibility/opentracing.md:51]: Makes sense, will relax it a little bit.

**Comment by carlosalberto** [comment]
@yurishkuro Updated to discourage users from using the OT Shim along the OTel API _only if_ baggage is used.

**Review by yurishkuro — APPROVED** [review]

**Review by jmacd — APPROVED** [review]
I believe the OTel-Go `bridge/opentracing` goes extensively out of it's way to make the behavior being specified here work, also (a) I'm not 100% certain of this statement because the code is difficult to follow, and (b) if I'm proven correct then we ought to simplify the code based on this spec change. I'll file an issue if you agree @carlosalberto.

**Pull request merged**